### PR TITLE
[CDAP-20277] Allow processing program status events in parallel for different run by sharding into N partitions

### PR DESCRIPTION
**Pull request by tivv**

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 1]
```diff
@@ -90,13 +93,82 @@
...
+    this.runRecordMonitorService = runRecordMonitorService;
+    List<Service> children = new ArrayList<>();
+    String topicPrefix = cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC);
+    int numPartitions = cConf.getInt(Constants.AppFabric.PROGRAM_STATUS_EVENT_NUM_PARTITIONS);
+    //Add bare one - we always listen to it
+    children.add(createChildService("program.status", topicPrefix));
+    //If number of partitions is more than 1 - create partitioned services
+    if (numPartitions > 1) {
+      IntStream.range(0, numPartitions).forEach(
+        i -> children.add(createChildService("program.status." + i, topicPrefix + i)));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java:143]: Do you have a subscribe that subscribe to the old topic too? It's for handling programs that starts before the upgrade. Maybe ut can simply be piggy backing on the subscriber for partition 0.

Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java:143]: Yes, it's line 139 right before the if

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStatePublisher.java [hunk 2]
```diff
@@ -49,14 +60,34 @@ public class MessagingProgramStatePublisher implements ProgramStatePublisher {
       .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
       .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec()).create();
   private final MessagingService messagingService;
-  private final TopicId topicId;
+  private final List<TopicId> topicIds;
   private final RetryStrategy retryStrategy;
 
-  public MessagingProgramStatePublisher(MessagingService messagingService,
-                                        TopicId topicId, RetryStrategy retryStrategy) {
+  @Inject
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStatePublisher.java:67]: it's kind of odd to have a public constructor with inject, as then it becomes difficult to tell when it's created through injection vs directly with the constructor. Looks like it's just injected in one new place, can we use the constructor there instead? Or alternatively switch this to package private and inject in the places it's manually constructed now?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java [hunk 3]
```diff
@@ -160,7 +173,7 @@ private TopicRelayer createTopicRelayer(CConfiguration cConf, String topic) {
     TopicId topicId = NamespaceId.SYSTEM.topic(topic);
 
     String programStatusTopic = cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC);
-    if (programStatusTopic.equals(topic)) {
+    if (topic.startsWith(programStatusTopic)) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java]: startsWith is a bit imprecise, can cause weird issues if somebody has changed the topic names for some reason. Would be safer to check if it's equal, or the topic prefix plus a number.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 4]
```diff
@@ -90,13 +93,82 @@
...
+  private final ProvisioningService provisioningService;
+  private final ProgramStateWriter programStateWriter;
+  private final TransactionRunner transactionRunner;
+  private final Store store;
+  private final RunRecordMonitorService runRecordMonitorService;
+  private final Service delegate;
+
+  @Inject
+  public ProgramNotificationSubscriberService(MessagingService messagingService, CConfiguration cConf,
+                                       MetricsCollectionService metricsCollectionService,
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java:117]: nit: indentation

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java [hunk 5]
```diff
@@ -90,13 +93,82 @@
...
+  private final ProgramLifecycleService programLifecycleService;
+  private final ProvisioningService provisioningService;
+  private final ProgramStateWriter programStateWriter;
+  private final TransactionRunner transactionRunner;
+  private final Store store;
+  private final RunRecordMonitorService runRecordMonitorService;
+  private final Service delegate;
+
+  @Inject
+  public ProgramNotificationSubscriberService(MessagingService messagingService, CConfiguration cConf,
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: this was package private before, does it need to be public?

**Review by albertshau — COMMENTED** [review]
minor comments

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStatePublisher.java (hunk 2 above)
Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStatePublisher.java:67]: Sure, let me inject it everywhere. It gives better isolation

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java (hunk 4 above)
Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java:117]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java (hunk 5 above)
Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java (hunk 3 above)
Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStatePublisher.java (hunk 2 above)
Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStatePublisher.java:67]: Done. Please review again

Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStatePublisher.java:67]: OK, I could not fully hide it due to io.cdap.cdap.app.guice.DistributedProgramContainerModule.ProgramStatePublisherProvider that needs a custom-configured ProgramStatePublisher that works through KubeDiscoveryService and not RemoteExecutionDiscoveryService

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/DistributedProgramContainerModule.java [hunk 6]
```diff
@@ -306,40 +308,39 @@ protected void configure() {
...
+  private static final class ProgramStatePublisherProvider implements Provider<ProgramStatePublisher> {
 
     private final CConfiguration cConf;
     private final MasterEnvironment masterEnv;
     private final InternalAuthenticator internalAuthenticator;
 
     @Inject
-    ProgramStateWriterProvider(CConfiguration cConf, MasterEnvironment masterEnv,
+    ProgramStatePublisherProvider(CConfiguration cConf, MasterEnvironment masterEnv,
                                InternalAuthenticator internalAuthenticator) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/DistributedProgramContainerModule.java]: nit: indentation

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStateWriter.java [hunk 7]
```diff
@@ -55,15 +50,7 @@ public final class MessagingProgramStateWriter implements ProgramStateWriter {
...
-  @Inject
-  public MessagingProgramStateWriter(CConfiguration cConf, MessagingService messagingService) {
-    this(new MessagingProgramStatePublisher(messagingService,
-                                            NamespaceId.SYSTEM.topic(cConf.get(
-                                              Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC)),
-                                            RetryStrategies.fromConfiguration(cConf, "system.program.state.")));
-  }
-
-  @VisibleForTesting
+  @Inject @VisibleForTesting
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStateWriter.java]: nit: we normally have annotations on different lines

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/ProgramStatePublisher.java [hunk 8]
```diff
@@ -26,6 +27,7 @@
 /**
  * Publishing program state messages and heartbeats
  */
+@ImplementedBy(MessagingProgramStatePublisher.class)
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/ProgramStatePublisher.java]: how come this is needed vs binding it in a guice module?

**Review by albertshau — COMMENTED** [review]
minor nits and one question

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/ProgramStatePublisher.java (hunk 8 above)
Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/ProgramStatePublisher.java]: Yes, I was considering the annotation vs module. Let me move to module to be consistent

Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/ProgramStatePublisher.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStateWriter.java (hunk 7 above)
Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/program/MessagingProgramStateWriter.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/DistributedProgramContainerModule.java (hunk 6 above)
Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/DistributedProgramContainerModule.java]: Done

**Review by albertshau — APPROVED** [review]
lgtm

**Pull request merged**